GowerMetric.py

```diff
@@ -0,0 +1,371 @@
+
+import numpy as np
+import sklearn.utils.validation
+from scipy.spatial.distance import pdist
+from scipy.cluster.hierarchy import (
+    linkage,
+    cophenet,
+    fcluster,
+)
+from scipy.optimize import minimize
+from sklearn.metrics import (
+    silhouette_score,
+)
+from sklearn.preprocessing import OrdinalEncoder, StandardScaler
+
+from utils import DataType
+
+
+class GowerMetric:
+    def __init__(
+        self,
+        dtypes: np.array,
+        ratio_scale_normalization="range",
+        weights=None,
+        precomputed_weights_file=None,
+    ):
+        self.dtypes = dtypes  # initialize with np.array of column data types
+        self.weights = weights
+        self.precomputed_weights_file = precomputed_weights_file
+        self.ranges_: List  # values of ranges in .ratio_scale() (iqr or traditional range)
+        self.h_: List  # h values in .ratio_scale()
+        self.n_features_in_: int
+        self.ratio_scale_normalization: str = ratio_scale_normalization
+
+        self._cat_nom_idx = [
+            True
+            if self.dtypes[i] == DataType.CATEGORICAL_NOMINAL
+            or self.dtypes[i] == DataType.BINARY_SYMMETRIC
+            else False
+            for i in range(len(self.dtypes))
+        ]
+        self._ratio_scale_idx = [
+            True if self.dtypes[i] == DataType.RATIO_SCALE else False
+            for i in range(len(self.dtypes))
+        ]
+
+        # Sums of columns of given type
+        self._cat_nom_num = sum(self._cat_nom_idx)
+        self._ratio_scale_num = sum(self._ratio_scale_idx)
+
+    def __call__(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+    ):
+        # TODO - no check for fit
+        # sklearn.utils.validation.check_is_fitted(self)
+        if self.n_features_in_ != len(vector_1) | len(vector_2):
+            assert len(vector_1) == len(vector_2)
+
+        # Distance for cat_nom and bin_sym columns
+        if self._cat_nom_num:
+            cat_nom_cols_1 = vector_1[self._cat_nom_idx]
+            cat_nom_cols_2 = vector_2[self._cat_nom_idx]
+            cat_nom_dist = 1.0 - (cat_nom_cols_1 == cat_nom_cols_2)
+
+            # Weighted variant - dot product with according self.weights columns
+            # Unweighted - a sum of distances of between all columns
+            if self.weights is not None:
+                cat_nom_dist = cat_nom_dist @ self.weights[self._cat_nom_idx]
+            else:
+                cat_nom_dist = cat_nom_dist.sum()
+        else:
+            cat_nom_dist = 0.0
+
+        # Distance for ratio_scale columns
+        if self._ratio_scale_num:
+            ratio_scale_cols_1 = vector_1[self._ratio_scale_idx]
+            ratio_scale_cols_2 = vector_2[self._ratio_scale_idx]
+            ratio_dist = np.abs(ratio_scale_cols_1 - ratio_scale_cols_2)
+
+            ones = ratio_dist >= self.ranges_[self._ratio_scale_idx]
+            zeros = ratio_dist <= self.h_[self._ratio_scale_idx]
```

Comment:
I don't get this part, or at least variable names. If I had to guess, if ratio is equal to or greater than range, we set it to maximum possible value 1, and if it's smaller or equal to appropriate `h`, we set it to minimal possible value 0? So you could just do:
```
ratio_dist = self.ranges_[self._ratio_scale_idx]
ratio_dist = np.clip(ratio_dist, -1, 1)
```
I omitted `self.h_` above, since it is used only for KDE normalization variant.